### ISSUE TITLE
Add support for data change notifications

### ISSUE THREAD

**Issue by RomanIakovlev** [clearlydefined/service#958]
The need to enable ClearlyDefined data replication has been discussed already (#386). I'd like to revive this discussion and propose to approach this problem as a 2 step process: 

1. Enable an externally visible notifications about changes in definitions, from both harvesting and curation. An event stream, if you'd like.
2. Implement data dump of all the existing definitions, with a selected set of attributes, to an external location (e.g. a file, a storage bucket, etc.).

This issue focuses on the first part, enabling the data change notifications. I propose to use a persistent web hook mechanism for this purpose. The list of requirements for ClearlyDefined service to implement this looks roughly like the following:

- Call an external web service when some data change happens.
- This call should keep on working after ClearlyDefined service restarts without any actions from the recipient (the persistent part of the definition above).
- Call URL should support secrets, e.g. auth credentials, without hardcoding them in the ClearlyDefined source code (obviously).
- There should be support for multiple web hooks, to have multiple events destinations.
- It is ok to use an ad-hoc, low-level approach for setting up new web hooks, e.g. via a PR to ClearlyDefined service. No web UI is required, at least initially.

Please feel free to add further requirements if they seem necessary, but having this seems to be sufficient for a start.

**Comment by jeffmcaffer** [comment]
see also #388

**Comment by RomanIakovlev** [comment]
Ok, so here's some more detailed proposal on how we could implement this.

Instead of web hooks as proposed in the original post, after a discussion with @jeffmcaffer I suggest we use a publicly available storage where we put all the newly created definitions. Here's how the storage might look like. Full disclosure, the layout of the storage is heavily inspired by the OpenStreetMap data replication approach.

## Storage layout

We structure that storage using time-based buckets. I propose the hourly buckets size to start with. Each bucket's name is an integer starting from 0 and incrementing.

To address the potential issue of having too many entries in the storage, we further structure the storage using top-level folders holding the buckets, and limiting the number of buckets in the folder to some large, but not overly large number, e.g. 1000 buckets per folder.

Inside each bucket, we keep a list of ClearlyDefined definitions in their raw form, as JSON files, named after the package coordinates. Additionally, inside each bucket we put a text file with both human- and machine-readable information about the bucket, e.g. creation timestamp and sequence number.

We may also consider using some tie-breaking mechanism to resolve multiple updates to the same package within an hour. For example, we could prefix the definition file names with a fixed-length unique string.

Here's how it might look like:

```none
000 \
  000 \
    _bucket-info.txt
    zBCscQyl/npm/npmjs/-/lodash/4.17.21.json
    2xqCz9jo/git/github/test-unit/test-unit/0eec725bbd0842bbfede1b4737766ff9ba4d178f.json
      ...
  001 \
    _bucket-info.txt
    hrg33jHG/gem/rubygems/-/test-unit/3.3.9.json
    ...
```

Feel free to browse e.g. [this](https://planet.maps.mail.ru/replication/hour/) OpenStreetMap replica mirror to get a better feeling for the proposed layout.

## How we put the definitions into the storage

In short, we could approach this by implementing a new kind of a definitions store (see [here](https://github.com/clearlydefined/service/tree/master/providers/stores) for a list of all of the stores) and use it as an additional [dispatch](https://github.com/clearlydefined/service/blob/master/providers/stores/dispatchDefinitionStore.js) target.

As soon as a new definition is created, either by a curation or a harvester callback, it goes through [definitionService._store](https://github.com/clearlydefined/service/blob/master/business/definitionService.js#L267) function, and that mechanism routes the new definition to the `dispatchDefinitionStore`. That store is configured using an environment variable `DEFINITION_STORE_PROVIDER` to route the stored objects to one or more of actual storage backends. The new definitions store, let's call it `publicDefinitionsStore`, will be added to the list of all actual storage backends handled by the `dispatchDefinitionStore`, just like this: `DEFINITION_STORE_PROVIDER: dispatch+azure+mongo+public`

## Storage ownership and technology choice

It's up for the discussion. We at GitHub are considering several possibilities, and I'd appreciate if we could get some input from the ClearlyDefined community as well. We have some options, like a publicly accessible cloud object storage (e.g. Azure Storage Account), or a dedicated server with fixed bandwidth. The biggest concerns here are the costs and abuse protection, as this would be a publicly available service, open to everyone on the Internet.

**Comment by DragosDanielBoia** [comment]
@RomanIakovlev about "How we put the definitions into the storage", we were thinking to use Azure Data Factory and use Azure Blob Storage (with share access signature)(this can be done incremental as well). Regarding the storage layout, maybe we can simplify (not sure we need the OpenStreetMap format) , dependents on how you want to query the files. Checking storage limits in azure, can't see any file number limits in a container (https://github.com/uglide/azure-content/blob/master/includes/azure-storage-limits.md , https://learn.microsoft.com/en-us/azure/storage/blobs/scalability-targets).

**Comment by RomanIakovlev** [comment]
Thanks for your feedback @DragosDanielBoia, it certainly brings several new things to the table. I didn't know you're about to move ClearlyDefined access behind auth, and that certainly changes the approach for data change notifications design.

I miss one thing from your proposal though. If all definition files are stored in a flat fashion without time buckets, how could the users know which files are new ones, and which they have already seen? It seems to me the users would be forced to re-download everything to get the updates? 

Or, in other words, I can't see the data change events in this setup. With the time buckets, users can track which buckets they have consumed, and don't have to re-process them the next time they pull changes.

I have one more concern regarding the merging of definitions and storing only one file per package version. But it's better to postpone that discussion until we sort out the storage layout, because the layout will have an impact on the merge decision.

**Comment by DragosDanielBoia** [comment]
@RomanIakovlev was thinking that we only keep last updated time ( and keep the time info only on that file, which is part of the blob properties when you save the file) and not the full history.  The user will have to remember which file with what time stamp they download last time. That's why was thinking about your querying scenarios.

**Comment by RomanIakovlev** [comment]
Could you please give some more details on how that index file approach would work, ideally with an example?

**Comment by DragosDanielBoia** [comment]
will do it tomorrow

**Comment by RomanIakovlev** [comment]
I suggest we mirror the complete definitions. Partly because it's probably easier to implement, partly because there might be different use cases in the future.

**Comment by DragosDanielBoia** [comment]
hi @RomanIakovlev  thanks for the patience, This is what we were thinking to do.

## Storage layout
1. We keep a flat list of files in a container
2. We have a folder in that container, changes, where we keep the changes. The file will be named date of the changes and contain which files ({type}.{provider}.{name}.{version}.json) were changed on that date (the field updated from the metadata). 

```
changes \
                2022-12-01
                2022-12-02
                ...

{type}.{provider}.{name}.{version}.json
...

```

With this approach we can have a point query first to check the changes and then to get them, preventing to iterate over the entire container.

## Things to consider when we implement
 - minimize the data we want to store; reasons
    - we want to minimize the egress,
    - some data in definitions is only used to compute the discovered license (by looking at document structure, I think we can ignore the highlighted fields) 
![image](https://user-images.githubusercontent.com/104837663/206212366-2a8296fe-675a-4d08-8cae-e5f4b7ccd510.png)
coordinates is the normalization of the key
- we should try to do this save data to storage (as a light data replication) outside of the clearlydefined system (the system is overloaded with too much code already for what it does), and if possible codeless (here we need to do some testing and explore the prices), which can be done in Azure (where the system resides)

**Comment by RomanIakovlev** [comment]
Thanks for the details @DragosDanielBoia! I think this looks good. I'm fine with dropping the `files` and `coordinates` parts. 

Regarding the updates latency. I'd like to understand when the `changes/<day>` files will be published. 

Here's an example. Now is 4pm CET, 7th of December. With this proposal, if I go into the `changes` folder right now, what would be the latest file I would find there?

From what I understand there's 2 options here. Option 1 is the latest file would be `2022-12-06`, or the day before today. This means as a user of this change notifications, my latency of getting updates would be at least 1 day. I think this is way too long, and it would be beneficial for the users to get the updates quicker.

Option 2 is the latest file would be `2022-12-07`, that is, today. This file is constantly updated as soon as any changes are published. Users can essentially poll the file with the desired rate and get the fresh changes. I think this is reasonable from the user perspective, but not sure if it's efficient.

As a compromise, I'd propose to go for the Option 1, but instead of days, store the hourly changes, so that the latency of getting updates is at least 1 hour. In other words, in the `changes` folder we'll have files like these:

```
changes \
  2022-12-07-00
  2022-12-07-01
  2022-12-07-02
  ...
  2022-12-07-23
  2022-12-08-00
  ...
  2022-12-08-23
```

As for whether to use the low-code approach to set this up, I think I'm fine with that, if this would allow us to get this done faster. By the way, I totally share your concern regarding the codebase complexity, so I think it would be beneficial to get this done on the side, not in the main codebase of ClearlyDefined main components (service and crawler). 

I'd just be wary of setting everything up using just click-ops (i.e. no code). Ideally this work would result in some sort of declarative description of this data synchronization process. And we should be able to check in this declarative description's source code into the ClearlyDefined repository, so that it can be repeated if necessary.

**Comment by qtomlinson** [comment]
@DragosDanielBoia @RomanIakovlev 
Storage of hourly changes + final definition json files sounds good.  A couple of points for consideration: 
- changes are essentially change logs for the hour: append only, file name (including path, see storage layout) and timestamp in each entry, no need to merge changes made for the same definition during the hour.
- definition file storage layout: instead of a flat file structure, organizing hierarchically by the current coordinates (via virtual folder in the blob storage) has some benefits:
```{type}/{provider}/{namespace}/{name}/version.json```	
  1. consistency with current REST api by including namespace.  
  2. intuitive, similar to maven repo and the current raw harvest data storage. 
  3. avoid any potential name collision because "." is used in component names.
  4. look for different versions of a component is fairly easy

  to provide extensibility for exposing other entities in the future, definition files can be put under definition, e.g. `definition/{type}/{provider}/{namespace}/{name}/version.json`

**Comment by DragosDanielBoia** [comment]
@qtomlinson this sounds good, thanks for the feedback.

1. regarding this one downside I see that if you want to see the changes hourly you will have to download the same file during that day, parse it and get  only the changes you need, and if the changes file is too big this will be a disadvantage, with hourly files, you go next hour and get only a small file to the previous hour then you are done with that file.

3. regarding the flat structure choice was to have prefix search, will have to check if blob storage supports prefix search when you have the subfolders structure, not sure about their implementation, if they have then will go with subfolders: prefix search will resolve your iv-th point.

**Comment by RomanIakovlev** [comment]
Thanks @qtomlinson for chiming in! I think the hierarchical coordinate-based layout sounds quite close to the naming scheme proposed by @DragosDanielBoia , so in that sense I don't mind either of approaches, let's just agree on one and go with it.

As for `changes` files being an append-only log, it looks like the main difference with what @DragosDanielBoia proposes is that the entries in `changes` files might be duplicated, in case the same coordinate changes more than once per hour. The definition files are presumably merged still, because otherwise the storage layout you propose doesn't work. Did I get this right?

In other words, the relationship between entries in `change` file and definitions is many-to-one, and and the user of change notifications would have to de-duplicate the `change` file entries to avoid re-downloading the same definition files. Does it make sense?

I presume it's easier to implement the append-only log approach (maybe it's more efficient as well?), at the expense of increased user complexity. I'd rather have it the other way around, making life easier for the users of this change notifications mechanism, even if the implementation gets (reasonably) more complex. But first and foremost I want to have this change implemented, and I think it's ok to have some trade-offs here in order to get this out of the door.

@DragosDanielBoia @qtomlinson regardless of what we agree upon, could we please have a mock / test version of data available for integration testing, ideally before the holidays start? I mean the actual blob storage account, with access granted, populated with some sample data, both definitions and changes.

**Comment by DragosDanielBoia** [comment]
@RomanIakovlev will try to get some sample data.

**Comment by qtomlinson** [comment]
@RomanIakovlev 
Yes. The definition json files are merged and only the final definition jsons are stored. The change file storage layout remains the same. The difference is the content of the change file. For example, change file 2022-12-07-00 can contain the following entries:

```
{ file: npm/npmjs/-/office-ui-fabric-react/7.176.2.json, timestamp: 2022-11-30T17:10:15.236534521Z }
{ file: git/github/angular-eslint/angular-eslint/f1e7704b7c0f57d7743fef8cd7500bc3103b9b5b.json, timestamp: 2022-11-30T17:13:50.133386195Z }
{ file: npm/npmjs/-/office-ui-fabric-react/7.176.2.json, timestamp: 2022-11-30T17:13:51.236536278Z }
```

If one accesses 2022-12-07-00 before the hour ends, one can then pick up from where downloading ends based on the timestamp. This assumes that the change file is being published constantly. If change file is only published after the hour ends, then storing only definition file names in change files (with duplications removed) makes sense.

**Comment by RomanIakovlev** [comment]
> This assumes that the change file is being published constantly. 
This is the key, now I understand what you mean. I thought the `changes` file is only published after the hour, and is published in its entirety at once.

If we want to go the way of constant updates for the last hour's file, I think it would be good to have some sort of mechanism to indicate the file is 'closed'. That is, it won't be appended to anymore. This is to let the users know the file won't change, and if they have finished processing it, they don't have to return to it again. We can achieve this for example by appending some fixed text entry at the end of the file. Something like `"done"` literal. 

But I have a feeling this is somewhat a brittle approach, for both implementers and users. I won't go into the details why I think it's brittle for the sake of brevity, but I think we should try and stick to the once-per-hour publishing, if possible. What do you think?

**Comment by qtomlinson** [comment]
@RomanIakovlev @DragosDanielBoia Is it worthwhile to add a way of schema versioning to the blob storage for future rework/expansion.

**Comment by RomanIakovlev** [comment]
@qtomlinson Maybe! What do you have in mind for that?

**Comment by qtomlinson** [comment]
How about a /metadata.json (for a hierarchical file structure) or metadata/metadata.json (easier to discover for a flat structure)? The file content includes storage layout/schema version and the reference to the latest change file: `{ version: '1.0.0', latest: 'changes/2022-12-07-00' }`?

**Comment by RomanIakovlev** [comment]
@qtomlinson I think it depends what kind of problem are we solving by adding the `metadata.json`. I think if we'd want to introduce a change in the storage layout, it's inevitable the clients would have to do some code changes. 

I'm not very keen on introducing and implementing a programmatic way of adjusting to new storage layout in the clients, at least not in the beginning, until such a need becomes obvious. And when it does become obvious, we'll be more experienced and better equipped to solve this problem, compared to where we are now.

I think we're also not painting ourselves in the corner by not introducing any mechanism like that right away. If we want to do some storage layout changes in the future, we can implement them in a new, separate storage container / account, which exists in parallel to the old one(-s). In other words, we can use storage container as a versioning mechanism. We'd probably have to support the old and new layouts in parallel for some time anyway, to accommodate the clients that can't update quickly.

The same goes for the `latest` change. I think it might be really useful if we have a problem with the listing performance. So it looks like an optimisation, the one we can safely add later.

**Comment by DragosDanielBoia** [comment]
@RomanIakovlev @qtomlinson take a look. I produced a sample data from the dev env.  For now since we don't have a process to share the container access keys, I gave public read only access on the blob level. So you will first have to read the index, https://clearlydefineddevbackup.blob.core.windows.net/definitions/changes/index, which contains all changes files -  is basically one query point to get started.

**Comment by RomanIakovlev** [comment]
Thanks @DragosDanielBoia! I've tried downloading the index, hourly change files, and definitions. I was able to access most of them, however several definitions gave me `404 Not Found` error. It's probably ok because it's only for testing purposes, just wanted to let you know.

Here are those not found entries:

```
maven/mavencentral/avalon-framework/avalon-framework/4.1.5.json
npm/npmjs/-/angular/1.3.15.json
npm/npmjs/@ant-design/icons/1.1.15.json
npm/npmjs/-/redie/0.1.0.json
npm/npmjs/-/rx/4.1.0.json
npm/npmjs/-/rx/2.3.24.json
go/golang/github.com%2fsigma/go-emacs/v0.0.1.json
```

**Comment by DragosDanielBoia** [comment]
@RomanIakovlev just picked random a few definitions to save, for index I updated all, ignore these for now. Won't be in prod.

**Comment by qtomlinson** [comment]
@DragosDanielBoia Awesome work to put the proposal into reality!  A couple of points for your consideration:
1. index file indeed provides a clear entry point to `changes` folder.  Should it be sorted by date time and perhaps append only blob can be used for optimization? This file will always grow, and as a single entry point, it can be potentially a performance concern in the future.

    a. Its content is similar to [List Blob](https://learn.microsoft.com/en-us/rest/api/storageservices/list-blobs?tabs=azure-ad) under `changes` folder, can [List Blob under a namespace](https://learn.microsoft.com/en-us/rest/api/storageservices/enumerating-blob-resources#Subheading4) be leveraged?  List blob has paging functionality (sorting seems to be missing?)

    b. The changes folder will have 8000+ files added each year.  To improve listing performance, changes folder can be potentially organized hierarchically by year (or year/month).  index file under each folder will then have its size capped.

2. change file (2022-03-19-05) as json (2022-03-19-05.json) for improved readability and future expansion? One consideration is to include url which is used to call get blob: `{ url: https://myaccount.blob.core.windows.net/definitions/go/golang/github.com%2fsigma/go-emacs/v0.0.1.json }`.  By including url, users are agnostic to our storage implementation.

3. leveraging blob index tags. Tags (e.g. last update date) can be added when updating the definitions and changes (See [Set Blob Tags (REST API)](https://learn.microsoft.com/en-us/rest/api/storageservices/set-blob-tags)).  User can potentially list blobs by specifying last update date and hour range, and then download them.  See [Manage and find Azure Blob data with blob index tags](https://learn.microsoft.com/en-us/azure/storage/blobs/storage-manage-find-blobs?tabs=azure-portal).  This can provide an additional way to search for changed blobs. This can be done in addition to files under `changes` directory which are easy to discover.

**Comment by DragosDanielBoia** [comment]
@qtomlinson for 1.a) we won't allow container list operation, this is very costly, that's why we force point queries all the time, to not enumerate( blob storage support this level of access, read only without the possibility to list blobs). Yes we can store in sorted order in index, good point, will do it in production.
for 1.b) blob storage implementation allows storing billions of files in container, without perf degradation ( which is not the case with OS disk management implementation)( we have tested this at scale).  The only limit in here is the capacity 500TB.

for 2. the change file is not a json, I don't see a reason,  see my point with listing above, no enumerator over container will be allowed. You start from index all the time and you keep on your side which changes you have downloaded, then you get the rest through point queries as well.

for 3. yes was thinking will add tags on the index file, so you can look at the changes having a second option, thx for the point as well.

**Comment by RomanIakovlev** [comment]
I have a proposal regarding the `definitions/changes/index` file format. I'd like to have entries formatted so that they keep their temporal order when sorted as strings. We can do so by adding leading zeroes to the month, day and hour components. 

The reason for this proposal is so that the user of the file can sort the entries and get only those, which are listed after the certain cut-off date and time. 

Speaking concretely, instead of formatting like this `2022-3-19-5` I propose something like this: `2022-03-19-05`.

Another solution to this problem is to ensure on the ClearlyDefined side the entries in `index` file are always sorted, so that user doesn't have to sort on their side. A sortable format would be nice to have nonetheless.

**Comment by DragosDanielBoia** [comment]
this '2022-03-19-05', and and have them in sortable order in the index as well. I think we can go with both.

**Comment by bduranc** [comment]
Hi All,

I am generally okay with what is proposed here. My only concern is leaving out the "files" section.

Now, I understand the files amount to a substantial amount of data. But the files section can be invaluable to compliance folks (and likely others too).

If we remove the files section, I fear what we'd end up with is a list of discovered licenses with no way to investigate where exactly those findings occur.

For me at least, I feel this takes away from the benefit of having such replication and notification ability.  

Best,
Brian

**Comment by DragosDanielBoia** [comment]
Hi @bduranc, I'll do some tests to see how much we increase the azure subscription costs, unfortunately the files section of the document is the most expensive in size to store. Currently we have some constraints regarding the increase in paying for the azure subscription. Even without the files you can write a tool to debug, to understand which files were taken in consideration for discovered license.

**Comment by DragosDanielBoia** [comment]
@RomanIakovlev I did all the updates for the index format like we discussed, I did another run and output all data from the dev environment, [index file](https://clearlydefineddevbackup.blob.core.windows.net/definitions/changes/index), when you have time can you check again? Next week, mid will star running this in prod. Will take a few days for the initial run, after which everything will be differential.

**Comment by bduranc** [comment]
> After we've started ingesting the data from production environment on scale, I've noticed there's a problem with Golang definitions. As far as I can tell, even though they're present in the time bucket files, the actual files are not present in the container. Should I open a new issue to investigate this?
> 
> Here's just a quick example from the recent data:
> 
> ```shell
> curl https://clearlydefinedprod.blob.core.windows.net/production-snapshots/changes/2023-04-26-18
> 
> ...
> go/golang/github.com%2feonasdan/tempus-dominus/v6.4.1+incompatible.json
> go/golang/github.com%2feonasdan/tempus-dominus/v6.2.8+incompatible.json
> 
> curl https://clearlydefinedprod.blob.core.windows.net/production-snapshots/go/golang/github.com%2feonasdan/tempus-dominus/v6.2.8+incompatible.json
> <?xml version="1.0" encoding="utf-8"?><Error><Code>BlobNotFound</Code><Message>The specified blob does not exist.
> RequestId:2bee2a6c-b01e-0058-4818-797064000000
> Time:2023-04-27T14:57:01.1640737Z</Message></Error>%
> ```

I also observe the same problem as @RomanIakovlev. Doesn't appear to be affecting any other package types.

**Comment by DragosDanielBoia** [comment]
@RomanIakovlev @bduranc will check this week and let you know.

**Comment by DragosDanielBoia** [comment]
found the problem is the double encoding, is not from the code, is from azure blob storage, when you use %2f you have to encode to %252f, so azure will match the path, the second url will match the path


https://clearlydefinedprod.blob.core.windows.net/production-snapshots/go/golang/github.com%2ftryghost/ghost/v4.3.2+incompatible.json

https://clearlydefinedprod.blob.core.windows.net/production-snapshots/go/golang/github.com%252ftryghost/ghost/v4.3.2%2Bincompatible.json
